### PR TITLE
Add support for a custom, per-instance signal handler

### PR DESCRIPTION
**Pull request by lostman**
@sunfishcode 

We have a use case which requires setting up a custom signal handler. `wasmtime` already sets up it's own signal handler to deal with traps and out of bounds memory errors (for `MemoryType::Static`). Care needs to be taken for the handlers not to interfere with each other.

The aim of this PR is to sketch out the API for a custom per-instance signal handler:

```rust
instance.borrow_mut().set_signal_handler(
    |signum, siginfo, context| {
        ...
    }
});
```

Users can supply their own function of the type:

```rust
Fn(libc::c_int, *const libc::siginfo_t, *const libc::c_void) -> bool
```

Where the `bool` return type indicates whether the signal has been handled. This allows dealing with specific signals while leaving the rest to the default `wasmtime` handler.

### CODE REVIEW

File: crates/api/src/instance.rs [hunk 1]
```diff
@@ -58,6 +58,13 @@ pub struct Instance {
 }
 
 impl Instance {
+    pub fn set_signal_handler<H>(&mut self, handler: H)
```
Comment (sunfishcode) [inline crates/api/src/instance.rs]: It'd be good to have a documentation comment here, explaining what this is for and what the requirements of the handler are. In particular, we should describe the use of the return value.

File: crates/interface-types/tests/custom_signal_handler.rs [hunk 2]
```diff
@@ -0,0 +1,273 @@
+use core::cell::Ref;
```
Comment (sunfishcode) [inline crates/interface-types/tests/custom_signal_handler.rs]: This is a good test, but it's not really specific to interface types. Would it makes sense to move it to wasmtime's top-level tests directory?

**Comment by lostman** [comment]
> Overall this looks good! A stack of instances is a good way to handle per-instance handlers, so that we don't need to be switching the actual signal handling function. This may need to evolve in the future to make it easier to apply the same handler to groups of instances, but I don't think we need to figure that out now.

Sounds good. I will fix the issues you've pointed out and add documentation.

**Comment by lostman** [comment]
@sunfishcode

I'm not sure what is the best way to handle OS specific parts of the API. Looking at `SignalHandlers.cpp`, we'll need three separate versions of `InstanceSignalHandler`: one for Windows, one for Mac, and one for Linux.

Perhaps something like this?

```
cfg_if::cfg_if! {
    if #[cfg(target_os = "linux")] {
        pub type SignalHandler = dyn Fn(libc::c_int, *const libc::siginfo_t,
*const libc::ucontext_t) -> bool;

        pub fn signal_handler_none(
            _signum: libc::c_int,
            _siginfo_ptr: *const libc::siginfo_t,
            _ucontext_ptr: *const libc::ucontext_t,
        ) -> bool {
            false
        }
    } else if #[cfg(target_os = "windows")] {
        pub type SignalHandler = dyn Fn(winapi::um::winnt::EXCEPTION_POINTERS)
-> bool;

        pub fn signal_handler_none(
            _exception_info: winapi::um::winnt::EXCEPTION_POINTERS
        ) -> bool {
            false
        }
    } else if #[cfg(target_os = "macos")] {
        // TODO
    }
}
```

> Would it makes sense to move this call into HandleTrap, after the CheckIfTrapAtAddress?

I don't think that would be desirable. `HandleTrap` looks OS independent and we want the `Instance` signal handler to have access to the details of the signal.

**Comment by sunfishcode** [comment]
I don't yet know what to suggest for the OS specific parts, but I'm working on learning more about this.

**Comment by sunfishcode** [comment]
> I don't think that would be desirable. `HandleTrap` looks OS independent and we want the `Instance` signal handler to have access to the details of the signal.

Ok, that makes sense. I think we can merge this, and proceed to get some experience with it.

**Pull request merged**